lib/spack/spack/installer.py

```diff
@@ -1550,6 +1618,7 @@ def install(self):
                 # Only terminate at this point if a single build request was
                 # made.
                 if task.explicit and single_explicit_spec:
+                    self._send_task_failure(task.request.pkg.spec)
```

Comment:
I'm worried in general about the shotgun mods to this file. I think there are too many concerns in installer.

Is it possible to raise a general type of exception for failures (i.e. wrap the exception if it's not a type you can catch all in one `except` block) and send the task failure someplace *outside* the core logic? I just think there is too much going on here and it'll become unmaintainable.

Reply:
This might be out of scope for the PR here - I just looked at this section again and it's covered in try/except/finally so I think we would need to rethink the entire structure. The only changes I added don't modify this structure (even if it's not great), they just insert the monitor to send an error in (the many!) places that can occur. I'm not familiar enough with this section to know how to best delete all these try/excepts in favor of one massive one. But if someone else familiar wants to walk it through with me, I could help! But a major refactor of this logic feels out of scope for this PR, which is already too big :)

Reply:
Yeah let's not rework the exception structure -- but see below w.r.t hooks.